client/me/purchases/billing-history/filter-transactions.ts

```diff
@@ -0,0 +1,84 @@
+import { isValueTruthy } from '@automattic/wpcom-checkout';
+import { getLocaleSlug } from 'i18n-calypso';
+import moment from 'moment';
+import {
+	BillingTransaction,
+	BillingTransactionUiState,
+} from 'calypso/state/billing-transactions/types';
+
+/**
+ * Utility function for formatting date for text search
+ */
+function formatDate( date: string ): string {
+	const localeSlug = getLocaleSlug();
+	return moment( date )
+		.locale( localeSlug ?? '' )
+		.format( 'll' );
+}
+
+/**
+ * Utility function extracting searchable strings from a single transaction
+ */
+function getSearchableStrings( transaction: BillingTransaction ): string[] {
+	const rootStrings: string[] = Object.values( transaction ).filter(
+		( value ) => typeof value === 'string'
+	);
+	const dateString: string | null = transaction.date ? formatDate( transaction.date ) : null;
+	const itemStrings: string[] = transaction.items.flatMap( ( item ) => Object.values( item ) );
+
+	return [ ...rootStrings, dateString, ...itemStrings ].filter( isValueTruthy );
+}
+
+/**
+ * Utility function to search the transactions by the provided searchQuery
+ */
+function searchTransactions(
+	transactions: BillingTransaction[],
+	searchQuery: string
+): BillingTransaction[] {
+	const needle = searchQuery.toLowerCase();
+
+	return transactions.filter( ( transaction: BillingTransaction ) =>
+		getSearchableStrings( transaction ).some( ( val ) => {
+			const haystack = val.toString().toLowerCase();
+			return haystack.includes( needle );
+		} )
+	);
+}
+
+export function filterTransactions(
```

Comment:
This file and its test were moved here from `getFilteredBillingTransactions` because they are only used here and don't need to be a selector.